keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,556 @@
+
+Because Kubernetes is a very technical solution on which lots of developers work on a daily basis we should not abandon completely developers centered approach for logging but instead add a way for Kubernetes users to choose between those two perspectives depending on the cluster purpose - if the given cluster is used for development or is it a production cluster. However this does not mean that the human-readable output format provided for developers needs to be exactly the same as the current klog format.
+
+Apart from being easy to ingest into logging services logs in order to be useful needs also to be searchable meaning that it should be easy for the user to find the most relevant information in them. Solutions like Elastic or BigQuery makes querying structured logs very easy. What they do not provide is the consistent schema to be used by those logs. Without it querying logs will be still possible but finding specific information might be much harder. Therefore apart from simply having some structure logs produced by applications should also follow a schema which would enforce consistent naming, consistent way of referencing the same objects and common set of metadata to be stored for each of those objects. With such schema in place building queries for logs becomes much easier.
+
+### Goals
+
+* Provide structured logging interface in K8s codebase
+* Standardize commonly used logging keys (e.g. referencing k8s object, http request fields)
+* Propose sound strategy for migration ensuring its completeness
+* Make implementation of context based solutions like logs correlation or tracing within Kubernetes easier.
+
+### Non-Goals
+
+* We are **not** changing logging pipeline in K8s
+* We are **not** standardizing some specific log messages
+* We are **not** tackling logging keys lifecycle (should be addressed by separate KEP when new API is close to GA)
+
+## Proposal
+
+This KEP proposes introducing a structured logging into Kubernetes control-plane components. As part of this proposal we will introduce a new logging API that will replace klog. We will introduce a Kubernetes logging schema which will provide a way to define a common set of logging keys and ensure their consistent usage. To provide immediate benefit for adapting a new API we will propose a new logging format that will simplify ingestion of structured data into third party logging solutions. As part of changes in kubernetes logging we propose to tightly couple usage of logging and golang context.
+
+### Logging Interface
+
+When choosing new structured logging API following criteria has been used:
+* Selected API should provide abstract structured logging interface so that it would be easy to inject different implementations for the same API.
+* Selected API should be minimalistic so that it should be easy to reimplement, wrap or to mock it.
```

Comment:
I have to be honest, the idea of having a logging API so complicated you need to mock it regularly is a little scary.

Reply:
Well, the point does not imply that this would be a regular task. It is enough that it needs to be done once or twice and the complexity of the API will make such task a way harder.
But you are true that the more often this is done the more important the simplicity of the API becomes.